core/src/main/java/org/apache/iceberg/InheritableMetadata.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.io.Serializable;
+
+class InheritableMetadata implements Serializable {
+
+  static InheritableMetadata fromManifest(ManifestFile manifest) {
+    return new InheritableMetadata(manifest.snapshotId());
+  }
+
+  private final Long snapshotId;
+
+  private InheritableMetadata(Long snapshotId) {
+    this.snapshotId = snapshotId;
+  }
+
+  ManifestEntry apply(ManifestEntry manifestEntry) {
```

Comment:
This is going to cause performance to degrade. We should set the snapshot ID on the existing `ManifestEntry` instead of copying it.

Reply:
Actually, that might not be a good idea. I'll expose a separate method.